### PR TITLE
[IOTDB-870] tags and attributes output

### PR DESCRIPTION
**Pull request by haimeiguo**
tags and attributes in result dataset are in json format

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/query/dataset/ShowTimeSeriesResult.java [hunk 1]
```diff
@@ -134,13 +151,25 @@ public static ShowTimeSeriesResult deserialize(ByteBuffer buffer) {
     result.encoding = ReadWriteIOUtils.readString(buffer);
     result.compressor = ReadWriteIOUtils.readString(buffer);
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/ShowTimeSeriesResult.java:152]: we have the enums of these strings, I suggest that we use the enums to serialize, u can find it in :
```
package org.apache.iotdb.tsfile.file.metadata.enums;
```

File: server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java [hunk 2]
```diff
@@ -127,12 +128,27 @@ public static void constructPathAndDataTypes(List<PartialPath> paths, List<TSDat
     paths.add(new PartialPath(COLUMN_TIMESERIES_COMPRESSION, false));
     dataTypes.add(TSDataType.TEXT);
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java:129]: duplicate code, how about:
```
  private static PartialPath[] resourcePaths = new PartialPath[]{new PartialPath(COLUMN_TIMESERIES, false), .....};
  private static TSDataType[] resourceTypes = new TSDataType[]{TSDataType.TEXT, .....};

  public static void constructPathAndDataTypes(List<PartialPath> paths, List<TSDataType> dataTypes, List<ShowTimeSeriesResult> timeseriesList) {
    Collections.addAll(paths,resourcePaths);
    Collections.addAll(dataTypes,resourceTypes);
    ......
  }
```

File: server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java [hunk 3]
```diff
@@ -127,12 +128,27 @@ public static void constructPathAndDataTypes(List<PartialPath> paths, List<TSDat
...
 
-    Set<String> tagAndAttributeName = new TreeSet<>();
+    //check if timeseries result has tag or attribute
+    boolean hasTag = false;
+    boolean hasAttribute = false;
     for (ShowTimeSeriesResult result : timeseriesList) {
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java:134]: I don't think there's any need to check,  it's ok to show `null || spaces`, like :
```
+--------------+-----+-------------+--------+--------+-----------+----+----------+
|    timeseries|alias|storage group|dataType|encoding|compression|tags|attributes|
+--------------+-----+-------------+--------+--------+-----------+----+----------+
|root.sg1.d1.s1| null|     root.sg1|   INT64|     RLE|     SNAPPY|null|      null|
|root.sg1.d1.s2| null|     root.sg1|   INT64|     RLE|     SNAPPY|null|      null|
|root.sg1.d1.s3| null|     root.sg1|   INT64|     RLE|     SNAPPY|null|      null|
+--------------+-----+-------------+--------+--------+-----------+----+----------+

```

**Review by liutaohua — COMMENTED** [review]
nice work , but I have some suggestions that I hope you'll take

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java [hunk 4]
```diff
@@ -994,14 +993,13 @@ private boolean match(PartialPath fullPath, String[] prefixNodes) {
...
-                ansString.right[3], ansString.right[4], Collections.emptyMap()));
+                ansString.right[3], ansString.right[4], Collections.emptyMap(), Collections.emptyMap()));
           } else {
             // has tags/attributes
             Pair<Map<String, String>, Map<String, String>> pair =
                 tagLogFile.read(config.getTagAttributeTotalSize(), tagFileOffset);
-            pair.left.putAll(pair.right);
             res.add(new ShowTimeSeriesResult(ansString.left.getFullPath(), ansString.right[0], ansString.right[1], ansString.right[2],
-                ansString.right[3], ansString.right[4], pair.left));
+                ansString.right[3], ansString.right[4], pair.left, pair.right));
```
[the remark this answers is not in the document]
Comment (haimeiguo) [reply server/src/main/java/org/apache/iotdb/db/metadata/MManager.java]: edited. thank you :)

File: server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java [hunk 5]
```diff
@@ -152,10 +142,11 @@ public static QueryDataSet getQueryDataSet(List<ShowTimeSeriesResult> timeseries
       updateRecord(record, result.getName());
       updateRecord(record, result.getAlias());
       updateRecord(record, result.getSgName());
-      updateRecord(record, result.getDataType());
-      updateRecord(record, result.getEncoding());
-      updateRecord(record, result.getCompressor());
-      updateRecord(record, result.getTagAndAttribute(), paths);
+      updateRecord(record, result.getDataType().toString());
+      updateRecord(record, result.getEncoding().toString());
+      updateRecord(record, result.getCompressor().toString());
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java:147]: i think that too many type cast , how about used only in `serialize/deserialize`.

File: server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java [hunk 6]
```diff
@@ -166,26 +157,28 @@ public static QueryDataSet getQueryDataSet(List<ShowTimeSeriesResult> timeseries
...
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java:168]: same

File: server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java [hunk 7]
```diff
@@ -112,36 +122,16 @@ public static void filterQueryDataSource(QueryDataSource queryDataSource,
...
+    Collections.addAll(paths, resourcePaths);
+    Collections.addAll(dataTypes, resourceTypes);
   }
 
   public static QueryDataSet getQueryDataSet(List<ShowTimeSeriesResult> timeseriesList,
       ShowTimeSeriesPlan showTimeSeriesPlan, QueryContext context) {
     List<PartialPath> paths = new ArrayList<>();
     List<TSDataType> dataTypes = new ArrayList<>();
-    constructPathAndDataTypes(paths, dataTypes, timeseriesList);
+    constructPathAndDataTypes(paths, dataTypes);
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java:134]: I think that `constructPath` and `updateRecord` behavior belongs to `Dataset`, how about move it to `ShowTimeseriesDataSet ` class?

and next line ,  the `showTimeSeriesPlan` already exists in dataset...


@qiaojialin  what do you think?

**Comment by liutaohua** [comment]
Don't forget merge _`main`_ branch

File: server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java (hunk 7 above)
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/utils/QueryUtils.java:134]: It seems that the constructPathAndDataTypes is only used for ShowTimeseries, so moving it to ShowTimeseriesDataSet is ok.

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 8]
```diff
@@ -566,7 +567,8 @@ private QueryDataSet processShowStorageGroup(ShowStorageGroupPlan showStorageGro
   private QueryDataSet processShowTimeseries(ShowTimeSeriesPlan showTimeSeriesPlan,
       QueryContext context) throws MetadataException {
     List<ShowTimeSeriesResult> timeseriesList = showTimeseries(showTimeSeriesPlan, context);
-    return QueryUtils.getQueryDataSet(timeseriesList, showTimeSeriesPlan, context);
+    ShowTimeseriesDataSet showTimeSeriesDataSet = new ShowTimeseriesDataSet(showTimeSeriesPlan, context);
+    return showTimeSeriesDataSet.getQueryDataSet(timeseriesList);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:571]: ShowTimeseriesData should play as a query processor. 
It accepts a query plan and generates results.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/ShowTimeseriesDataSet.java [hunk 9]
```diff
@@ -19,44 +19,123 @@
...
+      return;
+    }
+    Field field = new Field(TSDataType.TEXT);
+    field.setBinaryV(new Binary(s));
+    record.addField(field);
   }
 
   @Override
   protected boolean hasNextWithoutConstraint() throws IOException {
+    if (!hasSetRecord) {
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/ShowTimeseriesDataSet.java:135]: i think that this code can be optimizing:
1. Take the data directly from the `MManager` and initialize it in the `constructor`
2. if the total amount of data is less than the `fetchSize`, which means there is no more data, reducing the access to the `MManager`.

> when hasLimit = false, limit was means fetchSize

3. if `index == result.size()== fetchSize`, get data from `MManager` again.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/ShowTimeseriesDataSet.java [hunk 10]
```diff
@@ -19,44 +19,123 @@
...
+
+  private void updateRecord(RowRecord record, Map<String, String> map) {
+    String text = map.entrySet().stream()
+        .map(e -> "\"" + e.getKey() + "\"" + ":" + "\"" + e.getValue() + "\"")
+        .collect(Collectors.joining(","));
+
+    updateRecord(record, text.length() == 0 ? null : "{" + text + "}");
+  }
+
+  private void updateRecord(RowRecord record, String s) {
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/ShowTimeseriesDataSet.java:123]: `putRecord` method should also be modified to `private`

**Review by liutaohua — COMMENTED** [review]
victory is in sight

**Comment by liutaohua** [comment]
LGTM

**Review by liutaohua — APPROVED** [review]

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConstant.java [hunk 11]
```diff
@@ -70,6 +70,8 @@ private IoTDBConstant() {
   public static final String COLUMN_DEVICES = "devices";
   public static final String COLUMN_COLUMN = "column";
   public static final String COLUMN_COUNT = "count";
+  public static final String COLUMN_TAG = "tags";
+  public static final String COLUMN_ATTRIBUTE = "attributes";
```
Comment (HTHou) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConstant.java:74]:
```suggestion
  public static final String COLUMN_TAGS = "tags";
  public static final String COLUMN_ATTRIBUTES = "attributes";
```

**Review by qiaojialin — APPROVED** [review]
Good work!

**Pull request merged**